rtcopy/rtcpd.man
.\" Copyright (C) 2000-2004 by CERN IT-ADC
.\" All rights reserved
.\"
.nh
.TH RTCPD "1castor" "$Date: 2009/08/18 09:43:01 $" CASTOR "RTCOPY daemon"
.SH NAME
rtcpd \- CASTOR remote tape copy daemon
.SH SYNOPSIS
.B rtcpd
.SH DESCRIPTION
\fBrtcpd\fP is the CASTOR tape mover. It runs as a daemon on each tape
server known to the CASTOR volume and queue manager (VDQM). An RTCOPY
client application sends a volume request to VDQM using the \fBrtcpc(3)\fP
routine. The volume request consists of the visual ID of the volume (VID),
the access mode (read/write), the address of the client (host + port number),
the client login name and uid/gid. Once the request is eligible to run
the VDQM server sends the volume request to the "rtcopy" service port on
the selected tape server and passes the tape volume request to it. The
\fBrtcpd\fP daemon listens on the "rtcopy" service port and forks a subprocess
upon each new connection. The child process receives the volume request and
contacts the client to receive the full tape copy request.
.PP
If the tape copy request is correctly constructed, the \fBrtcpd\fP child
process calls the \fBCtape\fP API to mount and position the tape. It also
allocates large memory buffers (by default 128MB) for the data transfer.
If it is a tape write request and the tape blocksize has been specified,
the \fBrtcpd\fP child process will also start to fill the memory buffers
with input disk file data, while it is waiting for the tape to be mounted and
positioned. \fBrtcpd\fP is highly multi-threaded and able to process
disk+network I/O in parallel with the tape I/O.
.PP
To access the tapes \fBrtcpd\fP uses the character devices created by the tape daemon in the /var/lib/castor/rtcopy directory. This directory is created by \fBrtcpd\fP at startup time, and must not be removed.

.SH ENVIRONMENT
The \fBrtcpd\fP daemon does not take any arguments. A set of configuration
variables can be used to change default settings of \fBrtcpd\fP run parameters.
New values of the configuration variables can be set either statically
through environment variables when starting the \fBrtcpd\fP daemon, or
dynamically (taken into account by all following request) by modifying the
local CASTOR configuration file "/etc/castor/castor.conf".
.PP
The following configuration variables are available:
.PP
.B RTCOPYD_BUFSZ
\- Size of individual memory buffers (default=5MB).
.PP
.B RTCOPYD_NB_BUFS
\- Number of individual memory buffers (default=32). Note that the
default amount of physical memory required per RTCOPY request is 128MB
(32 times 4MB). For a tape server with 3 drives the minimum amount of
physical memory required for RTCOPY is 384MB (3 times 128MB). The two
parameters \fBRTCOPYD_BUFSZ\fP and \fBRTCOPYD_NB_BUFS\fP can be used
to change the default memory usage.
.PP
.B RTCOPYD_CKSUMALG
\- The name of the default checksum algorithm to be used for the segment
checksums, if it is not specified by the client. The supported values for
\fBRTCOPYD_CKSUMALG\fP are currently "adler32", "crc32" or "none". If the
configuration is not specified, the "adler32" is taken as the default.
.PP
.B RTCOPY_DEBUG
\- Turn on printing of debug messages for each I/O operation to/from the
memory buffers. \fBRTCOPY_LOGLEVEL\fP should be set to 7.
.PP
.B RTCOPY_LOGLEVEL
\- Set logging level. Normal level (LOG_INFO) is 6 and debug level (LOG_DEBUG)
is 7.
.PP
.B RTCOPYD_MAX_TPRETRY
\- Set the maximum number of retries on retry-able tape operation errors
(mount/position). Default is 2.
.PP
.B RTCOPYD_MAX_CPRETRY
\- Set the maximum number of retries on retry-able tape copy errors.
Default is 2.
.PP
.B RTCOPY_PORT
\- Set the RTCOPY listen port number. Default is 5003.
.PP
.B RTCPD_THREAD_POOL
\- Set the disk I/O thread pool size. This is the number of disk I/O transfers
which can run in parallel. Default is 3.
.PP
.B RTCPD_PILEUP_WAIT_TIME
\- Protect the tape writing from very slow disk IO streams. This option allows
for suspening the tape writing for the given time if there are no buffers
available. The intention is to avoid that the tape stops and starts at
every buffer. The wait time should be configured to match the drive speed
and the configured size of the rtcpd memory buffers.
.PP
.B RTCOPYD_SELF_MONITOR
\- Switch on the RTCOPY self monitor thread. The RTCOPY self monitoring thread
is a mechanism to make sure that RTCOPY server processes always exit from
blocking calls. The self monitoring time limits for various operations (e.g.
tape mounts, tape position, disk file open, etc.) can be set through a
dedicated set of configuration variables:
.PP
.B RTCOPYD_MOUNT_TIME
\- is the self-monitor timeout to mount a volume. Default is 3600 seconds.
.PP
.B RTCOPYD_POSITION_TIME
\- is the self-monitor timeout to position to a tape file.
Default is 3600 seconds.
.PP
.B RTCOPYD_OPEN_TIME
\- is the self-monitor timeout open a file (disk or tape).
Default is 300 seconds.
.PP
.B RTCOPYD_READ_TIME
\- is the self-monitor timeout for a single read operation on a file
(disk or tape). Default is 300 seconds.
.PP
.B RTCOPYD_WRITE_TIME
\- is the self-monitor timeout for a single write operation on a file
(disk or tape). Default is 300 seconds.
.PP
.B RTCOPYD_CLOSE_TIME
\- is the self-monitor timeout to close a file (disk or tape).
Default is 300 seconds
.PP
.B RTCOPYD_STAGEUPDC_TIME
\- is the self-monitor timeout for a single update of a stager catalogue.
Default is 3600 seconds.
.PP
Each one of the listed environment variables can also be set through a
corresponding CASTOR configuration parameter pair in "/etc/castor/castor.conf".
The CASTOR configuration parameters are constructed such that the part
before the first underscore ("_") sets the configuration parameter "category"
and the remaining part sets the configuration parameter "name".
For instance: the environment variable \fBRTCOPYD_NB_BUFS\fP corresponds to
a "/etc/castor/castor.conf" entry "\fBRTCOPYD   NB_BUFS\fP". The environment variables
take precedence over "/etc/castor/castor.conf" configuration parameters.

There are a few CASTOR configuration parameters for which there is no
corresponding environment variable:
.PP
.B RTCOPYD <dgn>_ERRACTION
\- modifies the default error action for tape media errors for the specified
device group. For instance: \fBRTCOPYD SD3R_ERRACTION\fP sets the error
action for drives in the SD3R device group. The value has to be a decimal
number representing a combination of RTCOPY severity flags, see \fBrtcpc(3)\fP.
As an example: \fBRTCOPYD SD3R_ERRACTION 2056\fP forces a immediate
failure with \fBRTCP_NORETRY\fP severity in case of tape media errors
(2056 == \fBRTCP_FAILED|RTCP_NORETRY\fP).
.PP
.B RTCOPYD REUSE_MOUNT
\- Set to TRUE (case insensitive) if the rtcpd daemon should request the vdqmd
daemon to try and reuse tape mounts across requests.  Any value other than TRUE
is interpretted as FALSE.  The default value of this parameter is TRUE.  This
parameter is always read just before a tape is unmounted, therefore the rtcpd
daemon does not need to be restarted in order for a new value for this
parameter to take affect.

.SH FILES
.TP 1.2i
.B /etc/castor/castor.conf
CASTOR configuration file
.TP
.B /var/lib/castor/rtcpd
Directory where the tape daemon creates the character devices
.TP
.B /var/log/castor/rtcpd.log
RTCOPY server log file

.SH SEE ALSO
.BI rtcpc(3)

.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


rtcopy/rtcpd.c
/*
 * $Id: rtcpd.c,v 1.8 2009/08/18 09:43:01 waldron Exp $
 *
 * Copyright (C) 1999-2004 by CERN IT
 * All rights reserved
 */

/*
 * rtcpd.c - Remote Tape Copy Daemon main program.
 */

#include <stdlib.h>
#include <time.h>
#include <string.h>
#include <stdio.h>
#include <sys/types.h>                  /* Standard data types          */
#include <netdb.h>                      /* Network "data base"          */
#include <sys/socket.h>                 /* Socket interface             */
#include <netinet/in.h>                 /* Internet data types          */
#include <signal.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <errno.h>
#include <patchlevel.h>
#include <Castor_limits.h>
#include <log.h>
#include <unistd.h>
#include <net.h>
#include <osdep.h>
#include <trace.h>
#include <serrno.h>
#include <Cgetopt.h>
#include <Cinit.h>
#include <Cuuid.h>
#include <rtcp_constants.h>
#include <rtcp.h>
#include <rtcp_server.h>
#include "tplogger_api.h"

extern char *rtcpd_logfile;
extern int use_port;
extern int Debug;
extern int loglevel;
extern int rtcp_InitLog (char *, FILE *, FILE *, int *);
extern int Cinitdaemon (char *, void (*)(int));

char *getconfent();

static int ChdirWorkdir() {
    char *workdir = "/var/lib/castor/rtcpd", *p;
    struct stat st;
    mode_t save_mask;
    int rc;

    if ( ((p = getenv("RTCOPY_WORKDIR")) != NULL) ||
         ((p = getconfent("RTCOPY","WORKDIR",0)) != NULL) ) {
        workdir = strdup(p);
    }
    /*
     * Check and create if necessary the RTCOPY working directory
     */
    if ( stat(workdir,&st) == -1 ) {
        if ( errno == ENOENT ) {
            save_mask = umask(0);
            rc = mkdir(workdir,S_IRUSR|S_IWUSR|S_IXUSR|S_IRGRP|S_IWGRP|
S_IXGRP|S_IROTH|S_IWOTH|S_IXOTH);
            if ( rc == -1 ) {
                    rtcp_log(LOG_ERR,"Cannot create directory %s: %s\n",
                                      workdir,sstrerror(errno));
                    tl_rtcpd.tl_log( &tl_rtcpd, 3, 4,
                                     "func"   , TL_MSG_PARAM_STR, "ChdirWorkdir",
                                     "Message", TL_MSG_PARAM_STR, "Cannot create directory",
                                     "Workdir", TL_MSG_PARAM_STR, workdir,
                                     "Error"  , TL_MSG_PARAM_STR, sstrerror(errno) );
            } else {
                    rtcp_log(LOG_INFO,"Created directory %s\n",workdir);
                    tl_rtcpd.tl_log( &tl_rtcpd, 10, 3,
                                     "func"   , TL_MSG_PARAM_STR, "ChdirWorkdir",
                                     "Message", TL_MSG_PARAM_STR, "Created directory",
                                     "Workdir", TL_MSG_PARAM_STR, workdir );
            }
            umask(save_mask);
        }
    } else if ( !S_ISDIR(st.st_mode) ) {
        rtcp_log(LOG_ERR,"%s should be a directory !\n",workdir);
        tl_rtcpd.tl_log( &tl_rtcpd, 3, 3,
                         "func"   , TL_MSG_PARAM_STR, "ChdirWorkdir",
                         "Message", TL_MSG_PARAM_STR, "Should be a directory",
                         "Workdir", TL_MSG_PARAM_STR, workdir );
    }
    if ( chdir(workdir) == -1 ) {
        rtcp_log(LOG_ERR,"chdir(%s): %s\n",workdir,sstrerror(errno));
        tl_rtcpd.tl_log( &tl_rtcpd, 3, 4,
                         "func"   , TL_MSG_PARAM_STR, "ChdirWorkdir",
                         "Message", TL_MSG_PARAM_STR, "chdir",
                         "Workdir", TL_MSG_PARAM_STR, workdir,
                         "Error"  , TL_MSG_PARAM_STR, sstrerror(errno) );
        return(-1);
    }
    return(0);
}

int rtcpd_main() {
    int rc;
    int pid;
    char workdir[] = "/var/lib/castor/rtcpd";
    char tpserver[CA_MAXHOSTNAMELEN+1];
    int *listen_socket = NULL;
    int *request_socket = NULL;
    int accept_socket = -1;

    signal(SIGPIPE,SIG_IGN);
    /*
     * Ignoring SIGXFSZ signals before logging anything
     */
    signal(SIGXFSZ,SIG_IGN);

    rtcp_InitLog(NULL,NULL,NULL,NULL);

    /* init the tplogger interface */
    {
            mode_t save_mask;
            char *p;

            save_mask = umask(0);

            p = getconfent ("TAPE", "TPLOGGER", 0);
            if (p && (0 == strcasecmp(p, "SYSLOG"))) {
                    tl_init_handle( &tl_rtcpd, "syslog" );
            } else {
                    tl_init_handle( &tl_rtcpd, "dlf" );
            }
            tl_rtcpd.tl_init( &tl_rtcpd, 1 );

            tl_rtcpd.tl_log( &tl_rtcpd, 9, 2,
                             "func"   , TL_MSG_PARAM_STR, "rtcpd_main",
                             "Message", TL_MSG_PARAM_STR, "Logging initialized" );
            umask(save_mask);
    }
    request_socket = (int *)calloc(1,sizeof(int));
    if ( request_socket == NULL ) {
            rtcp_log(LOG_ERR,"main() calloc(int): %s\n",sstrerror(errno));
            tl_rtcpd.tl_log( &tl_rtcpd, 3, 3,
                             "func"   , TL_MSG_PARAM_STR, "rtcpd_main",
                             "Message", TL_MSG_PARAM_STR, "main() calloc (int)",
                             "Error"  , TL_MSG_PARAM_STR, sstrerror(errno) );
            return(1);
    }

    rc = rtcpd_InitNW(&listen_socket);
    if ( rc == -1 ) {
            rtcp_log(LOG_ERR,"main() rtcpd_InitNW(): %s\n",sstrerror(serrno));
            tl_rtcpd.tl_log( &tl_rtcpd, 3, 3,
                             "func"   , TL_MSG_PARAM_STR, "rtcpd_main",
                             "Message", TL_MSG_PARAM_STR, "main() rtcpd_InitNW()",
                             "Error"  , TL_MSG_PARAM_STR, sstrerror(serrno) );
            free(request_socket);
            return(1);
    }

    /*
     * Go to working directory
     */
    rc = ChdirWorkdir();
    if ( rc == -1 ) {
        rtcp_log(LOG_ERR,"main() ChdirWorkdir(): %s\n",sstrerror(errno));
        tl_rtcpd.tl_log( &tl_rtcpd, 3, 3,
                         "func"   , TL_MSG_PARAM_STR, "rtcpd_main",
                         "Message", TL_MSG_PARAM_STR, "main() ChdirWorkdir()",
                         "Error"  , TL_MSG_PARAM_STR, sstrerror(errno) );
        free(request_socket);
        return(1);
    }

    gethostname(tpserver,CA_MAXHOSTNAMELEN);
#if defined(__DATE__) && defined (__TIME__)
    rtcp_log(LOG_INFO,"Rtcopy server generated at %s %s. Started in %s:%s\n",
             __DATE__,__TIME__,tpserver,workdir);
    tl_rtcpd.tl_log( &tl_rtcpd,  11, 5,
                     "Message",  TL_MSG_PARAM_STR, "Rtcopy server generated",
                     "date",     TL_MSG_PARAM_STR, __DATE__,
                     "time",     TL_MSG_PARAM_STR, __TIME__,
                     "tpserver", TL_MSG_PARAM_STR, tpserver,
                     "workdir",  TL_MSG_PARAM_STR, workdir );
#endif /* __DATE__ && __TIME__ */

    rc = 0;
    for (;;) {
        rc = rtcp_Listen(*listen_socket,&accept_socket,-1, RTCP_ACCEPT_FROM_CLIENT);
        if ( rc == -1 ) {
            rtcp_log(LOG_ERR,"main() rtcp_Listen(): %s\n",
                sstrerror(serrno));
            tl_rtcpd.tl_log( &tl_rtcpd, 3, 3,
                             "func"   , TL_MSG_PARAM_STR, "rtcpd_main",
                             "Message", TL_MSG_PARAM_STR, "main() rtcp_Listen()",
                             "Error"  , TL_MSG_PARAM_STR, sstrerror(serrno) );
            continue;
        }

        pid = (int)fork();
        if ( pid == -1 ) {
            rtcp_log(LOG_ERR,"main() failed to fork(): %s\n",
                     sstrerror(errno));
            tl_rtcpd.tl_log( &tl_rtcpd, 3, 3,
                             "func"   , TL_MSG_PARAM_STR, "rtcpd_main",
                             "Message", TL_MSG_PARAM_STR, "main() failed to fork()",
                             "Error"  , TL_MSG_PARAM_STR, sstrerror(errno) );
            continue;
        }
        if ( pid != 0 ) {
            close(accept_socket);
            continue; /* Parent */
        }

        /*
         * Child
         */
        signal(SIGPIPE,SIG_IGN);
        /*
         * Ignoring SIGXFSZ signals before logging anything
         */
        signal(SIGXFSZ,SIG_IGN);
        close(*listen_socket);
        free(listen_socket);
        listen_socket = NULL;
        *request_socket = accept_socket;
        rc = rtcpd_MainCntl(request_socket);
        break;
    }
    (void)rtcp_CleanUp(&listen_socket,rc);
    if ( request_socket != NULL ) free(request_socket);

    tl_rtcpd.tl_exit( &tl_rtcpd, 0 );

    return(0);
}

int main(int argc, char *argv[]) {
    int c;

    Coptind = 1;
    Copterr = 1;

    while ( (c = Cgetopt(argc, argv, "df:p:")) != -1 ) {
        switch (c) {
        case 'd':
            Debug = TRUE;
            loglevel = LOG_DEBUG;
            NOTRACE;
            break;
        case 'f':
            rtcpd_logfile = strdup(Coptarg);
            if ( rtcpd_logfile == NULL ) {
                fprintf(stderr,"main() strdup(): %s\n",
                    strerror(errno));
                return(1);
            }
            break;
        case 'p':
            use_port = atoi(Coptarg);
            break;
        default:
            break;
        }
    }

    if ( Debug == TRUE ) exit(rtcpd_main());

    if ( Cinitdaemon("rtcpd",SIG_IGN) == -1 ) exit(1);
    exit(rtcpd_main());
}
